### PR TITLE
Nathan permissions merge

### LINKED ISSUE #6: Set permissions for time log.

### PR DESCRIPTION
**Pull request by nathanah**
# Description
Fixes #6 (HIGH) Verify that permissions are implemented on the back-end and not just the front-end.
Permissions were split into front-end permissions and back-end permissions with some systems designed to accommodate this. This PR merges both permission sets into one so it's easier to make sure all permissions have relevant security checks implemented.

## Related PRS (if any):
This frontend PR is related to the #416 backend PR. 
https://github.com/OneCommunityGlobal/HGNRest/pull/416

## Main changes explained:
- I did a search and replace on all the front-end permissions to replace them with relevant back-end permissions.
- Deleted associatedPermissions.js and all references to it.
- Added Redux to hasPermission() calls to simplify arguments
- Memoized hasPermission() results

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Everything should work the same as before for every user.

## Screenshots or videos of changes:
Everything should look the same other than the code.

## Note:
If someone wants to help with unit testing, that would be appreciated. I'm considering it out of the scope of this PR though.

### CODE REVIEW

File: src/components/Projects/Projects.jsx [hunk 1]
```diff
@@ -159,8 +159,8 @@ export class Projects extends Component {
             onClick={this.toggleProjectInfoModal}
           />
           <Overview numberOfProjects={numberOfProjects} numberOfActive={numberOfActive} />
-          {hasPermission(role, 'addProject', roles, userPermissions) ? (
-            <AddProject addNewProject={this.addProject} />
+          {hasPermission(role, 'postProject', roles, userPermissions) ? (
+            <postProject addNewProject={this.postProject} />
```
Comment (cone-7) [inline src/components/Projects/Projects.jsx:163]: The name of a component must begin with an uppercase letter `<PostProject ...> `.

File: src/components/UserProfile/TeamsAndProjects/AddProjectPopup.jsx [hunk 2]
```diff
@@ -33,7 +33,7 @@ const AddProjectPopup = React.memo(props => {
       <ModalHeader toggle={closePopup}>Add Project </ModalHeader>
       <ModalBody style={{ textAlign: 'center' }}>
         <div className="input-group-prepend" style={{ marginBottom: '10px' }}>
-          <AddProjectsAutoComplete
+          <postProjectsAutoComplete
```
Comment (cone-7) [inline src/components/UserProfile/TeamsAndProjects/AddProjectPopup.jsx:36]: Same, component name `<PostProjectsAutoComplete >`

File: src/components/UserProfile/TeamsAndProjects/AddProjectPopup.jsx [hunk 3]
```diff
@@ -59,4 +59,4 @@ const AddProjectPopup = React.memo(props => {
     </Modal>
   );
 });
-export default AddProjectPopup;
+export default postProjectPopup;
```
Comment (cone-7) [inline src/components/UserProfile/TeamsAndProjects/AddProjectPopup.jsx:62]: Component name.

**Review by cone-7 — CHANGES_REQUESTED** [review]
We should follow the naming convention for components.

File: src/components/Projects/Projects.jsx (hunk 1 above)
Comment (KurtisIvey) [reply src/components/Projects/Projects.jsx:163]: I agree

Comment (nathanah) [reply src/components/Projects/Projects.jsx]: I changed 'post' back to 'Add' for all of these. Let me know if I missed any.

**Comment by yufeiStella** [comment]
As far as it is good on the local site.

File: src/components/Badge/BadgeReport.jsx [hunk 4]
```diff
@@ -303,7 +303,7 @@ const BadgeReport = props => {
                 <th style={{ width: '110px' }}>Modified</th>
                 <th style={{ width: '110px' }}>Earned Dates</th>
                 <th style={{ width: '90px' }}>Count</th>
-                {hasPermission(props.role, 'deleteOwnBadge', roles, props.permissionsUser) ? (
+                {hasPermission(props.role, 'deleteBadges', roles, props.permissionsUser) ? (
```
Comment (cone-7) [inline src/components/Badge/BadgeReport.jsx:306]: The function is being calling multiple times, I think it's better if you create a variable
```suggestion
                const hasDeleteBadgedPermission = hasPermission(props.role, 'deleteOwnBadge', roles, props.permissionsUser);
                ...
                {hasDeleteBadgedPermission ? (
```

**Review by winghojackyli — CHANGES_REQUESTED** [review]
Hi @nathanah , I've tested the PR and it works fine except for the tasks part. It pops out error and stop rendering the page. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108389005/79f8b624-5631-495e-8b17-13035212a1c1

**Comment by nathanah** [comment]
> Hi @nathanah , I've tested the PR and it works fine except for the tasks part. It pops out error and stop rendering the page.
> 
>  HGN.APP.-.Google.Chrome.2023-07-29.10-03-58.mp4

This was caused by another PR and has since been reverted.

**Review by winghojackyli — DISMISSED** [review]
The PR looks good and everything works as before.

**Review by ThapeloMasasa — COMMENTED** [review]
The code looks pretty code, good recognition for the use of proper camelCase. Everything works as intended

**Review by DavidC0126 — COMMENTED** [review]
Hi @nathanah,
I have tested your PR and everything works great except for create user. When I click user management -> create new user, the page became blank. (I tried to use different roles like Admin or Owner but they both behaved like this.)
blank_page

**Review by iTvX — CHANGES_REQUESTED** [review]
I have done my best to check everything, everything looks fine except for I can not found permission management under Other links tab. I think you should fix that

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/46791134/c132b499-be18-4ed0-8a40-bab7e4d9b08e

File: src/__tests__/Dashboard.test.js [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
 import '@testing-library/jest-dom/extend-expect';
 import mockState from './mockAdminState.js';
-import Dashboard from '../components/Dashboard';
+// import Dashboard from '../components/Dashboard';
 import thunk from 'redux-thunk';
 import configureStore from 'redux-mock-store';
-import { Route } from 'react-router-dom';
-import userEvent from '@testing-library/user-event';
+// import { Route } from 'react-router-dom';
+// import userEvent from '@testing-library/user-event';
```
Comment (Abiddy) [inline src/__tests__/Dashboard.test.js:10]: We could get rid of the comments once this is finalized for deployment.

**Review by Abiddy — DISMISSED** [review]
Everything looks good as far as I have checked. A lot of useful changes here. Nice.

**Review by winghojackyli — APPROVED** [review]
Hi @nathanah , I've tested this PR and everything looks good as the development branch. 
Owner, admin and volunteer accounts have been tested on this PR. Owner can see permission list but admin and volunteer cannot. I can also use owner and admin accounts to create new users, new teams and new projects and edit them. 
Great work!

**Review by KurtisIvey — CHANGES_REQUESTED** [review]
volunteer account functions as it should. Doesn't have access to different areas that it shouldn't.
admin & owner account seems to have the exact same permissions as the volunteers except for on the tasks section of the dashboard. I checked the redux state, and it seems they should be able to. However, I haven't gone into the code to try and diagnose why this isn't occurring since I'm not familiar with that area of the codebase.

admin should be able to pause people:
pr931 admin cant pause volunteer

admin: can't access reports or others tab
pr931 admin acct does not have access to reports   others

owner: can't access reports or others tab:
pr931 owner cant access reports or others tab

**Review by veronicacheng2 — APPROVED** [review]
Hi @nathanah, I have tested the PR using admin and owner account. There are no breaking changes and the app works as usual when I am navigating through different pages.

Logged in as admin

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95119686/660cb31e-8373-40c0-a8c2-3f79852bfe1e



Logged in as owner


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95119686/de56ce92-dd85-4d5e-96df-756193d6226e

**Review by KurtisIvey — APPROVED** [review]
Functions well in subsequent re-reviews with pull requesters guidance. Was able to clear up the reports and other links issue by running npm run build.  Great Job and thanks for your patience.

**Review by KurtisIvey — APPROVED** [review]
Fixes remained consistent after merge conflict resolved. Everything functions as it should

**Review by ss23046 — APPROVED** [review]
Hi @nathanah ,
I read all the comments and tested the PR. App runs normally. The above requested changes have been resolved. 

Permissions management tab is seen
Screen Shot 2023-08-19 at 4 14 24 PM

Able to navigate and access reports and other links(can create user too)
Screen Shot 2023-08-19 at 4 17 16 PM
Screen Shot 2023-08-19 at 4 16 22 PM

Admin is able to pause 
Screen Shot 2023-08-19 at 4 16 34 PM

**Pull request merged**